### ISSUE TITLE
time series with multiple units

### ISSUE THREAD

**Issue by calestyo** [grafana/grafana#79547]
Hey.

I've searched quite a lot whether this is possible right now, and there are indeed numerous community posts (also [brought it newly up](https://community.grafana.com/t/simple-way-for-multiple-y-axes-for-the-same-query-but-with-different-units/109020) myself), stackoverflow questions and the likes (especially about dual y-axis), but none seems to really provide a proper solution.

What  I want are ways to handle graphs, where a single time series may have multiple different units.

Consider for example a metric that shows a temperature or one that shows network traffic or disc IO.<br>
Common units for the former may be °C, K, or °F… and fort the latter it may be bits/s, bytes/s and the the whole things with different bases (1000 vs. 1024).

**Why is this needed**:
Right now it seems, one has to hardcode a single fixed unit, and this is it.<br>
If one wants to see the graph or gauge for multiple units, one needs multiple graphs or gauges.

For such metrics, a plotted graph would look the same, regardless of the chosen unit - only the numbers at the y-axis would change and the exact numbers displayed in the tool-tip.<br>
For gauges, the needle would at least point in the same direction (only the displayed number would be different).

Plotting the effectively same graph in a panel (only with different units), needlessly costs performance and just messes up the panel (e.g. the legend). For gauges, one just wastes a lot of space when one simply uses multiple gauges (each with another unit).

There are IMO many scenarios where one does want multiple units, best examples are those from above, for e.g. disk IO, I likely want both bytes(1000)/s and bytes(1024)/s... (bits/s are rather not used with disk IO, but with network IO)... and basically every other type of metric, where multiple units are commonly used.

**What would you like to be added**:

I'd want ways to handle such multiple units, though there may not be one golden solution for all cases.

Some ideas I had:
1. Using a 2nd y-axis for a single graph, like I tried to fake [here](https://community.grafana.com/t/simple-way-for-multiple-y-axes-for-the-same-query-but-with-different-units/109020/3?u=calestyo):
   ![image](https://github.com/grafana/grafana/assets/919933/1daf27c8-42cf-411f-979e-38e0c988fd24)
   The problem with that of course is: It's in reality all graphs duplicated (and all plotted twice, and twice in the legend).

   That should be made just work properly, i.e. truly multiple units for **one** graph only. And if the value is shown in the tool tip, it would indeed by nice if one could configure whether it's shown there for all units or just for some of them... but regardless of what one configures there: all the values under one label (with make fake solution above, there would be multiple labels).

2. (1) may be a nice way for some use cases, but especially when having more than just 2 units, or when panels contain also other graphs with completely unrelated units, (1) may get a bit messy.
   So my actually preferred solution, if I had to choose only one, was that one can use variables to select the used unit. So in the top of my dashboard I'd have some drop downs, where I can chose which unit I want to use now. And panels could refer to such variable and use that as unit.
   Problem here is of course, that the system would need to know how to convert, i.e. if the time series is in bits/s, how does it convert to bytes/s?

Any ideas?

Thanks,
Chris.

**Comment by leeoniya** [comment]
i'm a bit confused by the question.

is this not what you're looking for? (see field overrides configured in this panel)
https://play.grafana.org/d/000000016/1-time-series-graphs?orgId=1&viewPanel=540

what's the difference with your use case?

can you attach your panel (without any attempted work-arounds / duplications, so we can work from a clean initial state? https://grafana.com/docs/grafana/latest/troubleshooting/send-panel-to-grafana-support/

**Comment by calestyo** [comment]
> is this not what you're looking for? (see field overrides configured in this panel)
> https://play.grafana.org/d/000000016/1-time-series-graphs?orgId=1&viewPanel=540

No. AFAIU, this example is: 1 panel (with overall 3 units in the y-axes) and 3 graphs **but each of them with 1 unit**. So you have the blue graph, which has Watts, the yellow is Pascal and the green °Celsius.

What I want is that I have **multiple units for one graph**, e.g. you'd have a curve that shows the read data per second, and gives that in both: MB/s and MiB/s (that is: bases 1000 and 1024).

> can you attach your panel (without any attempted work-arounds / duplications, so we can work from a clean initial state? 

[debug-gar-ws-bl28.garching.physik.uni-muenchen.de-2023-12-15 00 42 45.json.txt](https://github.com/grafana/grafana/files/13679306/debug-gar-ws-bl28.garching.physik.uni-muenchen.de-2023-12-15.00.42.45.json.txt)
But that's really just the shamelessly stolen "Network Traffic Basic" panel from the [Node Full Exporter](https://grafana.com/grafana/dashboards/1860-node-exporter-full/) dashboard, with only a few modifications (making it work with a variable-configured datasource, make it repeated and use the hostname as title).

I has merely two queries, one for sent and one for received traffic... and uses per default bits/s.

What I'd want to be in the description of (1) above would e.g. be to have each of those two with bits/s **and** bytes/s.

And in (2) the idea would be that one can configure this via a variable, which I think would be practically more useful.

**Comment by leeoniya** [comment]
oh ok, just a secondary conversion / alternative scale axis. like https://leeoniya.github.io/uPlot/demos/dependent-scale.html

i agree it is useful in some cases, but would be difficult to do with Grafana's architecture, since y axes are dynamically generated from fields, and fields only have one assignable unit.

your workaround duplicate the field is probably the only way to do it. you can set 0 opacity and 0 line width on the dupe, and then add a field override `hideFrom > legend` and `hideFrom > tooltip`.

it's unlikely this will get worked on due to high difficulty and this is the first time i'm seeing a request for this in 3+ years.

**Comment by calestyo** [comment]
> i agree it is useful in some cases, but would be difficult to do with Grafana's architecture, since y axes are dynamically generated from fields, and fields only have one assignable unit.

Well, therefore my idea (2) - make the unit configurable via a variable, which might be much easier... and would perhaps anyway be much more what one wants in practise.

That alone would however not be enough, cause if my query gives me e.g. bytes, and the unit then makes bits, the value would need to be converted, so that Grafana could print proper axis descriptions and tooltips values.<br>
So I guess for that, one would need some factor, e.g. the query gives some base unit where the factor would effectively be `1`. If I now select bits/s, then the creator of the panel would need to somehow attach the factor `8` to the user-selected value, and modify to query to multiply the value by the factor.

**Comment by calestyo** [comment]
>  you can set 0 opacity and 0 line width on the dupe, and then add a field override hideFrom > legend and hideFrom > tooltip.

Thanks for that.

**Comment by leeoniya** [comment]
> What if one had a new query type, which simply uses the data from another query (without re-doing the query against the source or without plotting any graph), but allowing to multiply with some factor?

you can use the "Add field from caclulation" transformation to create new fields from other fields, so i think that will cover your case:

https://grafana.com/docs/grafana/latest/panels-visualizations/query-transform-data/transform-data/#add-field-from-calculation

> Operation - If you select Binary operation or Unary operation mode, then the Operation fields appear. These fields allow you to apply basic math operations on values in a single row from selected fields. You can also use numerical values for binary operations.

**Comment by leeoniya** [comment]
> Well, therefore my idea (2) - make the unit configurable via a variable, which might be much easier... and would perhaps anyway be much more what one wants in practice.

i don't think that one will be as easy as https://github.com/grafana/grafana/pull/78086, but maybe worth looking into at some point.

**Comment by calestyo** [comment]
> you can use the "Add field from caclulation" transformation to create new fields from other fields, so i think that will cover your case:

I tried that now, but I think it rather not works in a way to be useful for this issue.

When I "copy" the original query by adding a field from a calculation, and doing e.g. *8 on the original query... I will actually get two **differently scaled** graphs (though with the same "shape") in the panel, which is of course not what I want (I want only one graph, but with two values in units and/or 2 y-axes).

So I tried to simply disable visualisation of the transformation graph, via an Override, Series > Hide in area.<br>
But the problem here is, when I disable "Viz" for the graph, Tooltip also gets dysfunctional.<br>
Could that be a bug?

The 2nd y-axis also doens't really work:
![Screenshot from 2023-12-18 05-34-00](https://github.com/grafana/grafana/assets/919933/facce839-5baf-4268-91da-ea3bd4415d73)
If I just add one, I get the above, and the graph from the transformation is placed on another zero line.
I can enable "Centered Zero", but that's also not the same:
![Screenshot from 2023-12-18 05-35-57](https://github.com/grafana/grafana/assets/919933/d9de9b16-d2f6-458f-aa18-035300b4a5b3)
As you can see, both the positive and negative side of the y-axis are now scaled to the same maximum values, so I waste a lot of space. Also, for some reason, I see now both upper graphs (on still sees a bit of the green line, though it should in principle match).

**Comment by calestyo** [comment]
Zooming in shows it a bit better:
![Screenshot from 2023-12-18 05-48-00](https://github.com/grafana/grafana/assets/919933/1346e33b-d680-43e8-84d1-e016cc0a4a02)

On the upper side, there's the two graphs (centered to zero),... but they're somehow shifted.

**Comment by calestyo** [comment]
btw: Another case in which one IMO nearly always wants multiple units is everything where percentages are used.

When one has e.g. a graph like filesystem free space (in percent), then it would be quite convenient, if at the same time one could also see the actual numbers.

**Comment by calestyo** [comment]
Just because nothing happens, doesn't mean that the issue magically went away. IMO annoying bots like that merely send the message that any contributions in form of bug or feature requests are undesired.

**Comment by leeoniya** [comment]
i agree with you on the stalebot thing, but realistically [in this case] we have to prioritize work based on impact and effort. given that there have been zero upvotes or comments on this issue in a year, the impact is quite low - feel free to compare it to all other dataviz requests that exist in this repo with dozens or hundreds of upvotes.

the unit you select for a field isn't strictly a display unit, it's you telling us what unit the data is in, and the rendered axis ticks will actually adapt the displayed unit to be "nice", so 1000m becomes 1km, if you defined the unit in "meters", and 0.9km becomes 900m if you defined the unit as "kilometers".

you data cannot be in more that one unit. so what you're asking for is not just the ability to add more axes with units, but also conversion functions from the field/data unit to whatever additional axis you may want to configure. obviously you cannot convert kg -> C, so this now requires lookup tables for units that can be converted, and also bidirectional conversion functions for all such units. C -> K is simple addition, but C -> F is an expression, etc. this is all just the tip of what is sure to be an iceberg (decimals, log scaling, zero-alignment...)

while the underlying plotting lib does support multiple scales/axes that are derived from other scales ([demo](https://leeoniya.github.io/uPlot/demos/dependent-scale.html)), the effort of doing this for all units that Grafana supports goes beyond simply throwing together some UI to define additional axes for a field.

how much value is there in leaving this open with one person in 12mo asking, and no one to work on it?

> I've searched quite a lot whether this is possible right now, and there are indeed numerous community posts (also [brought it newly up](https://community.grafana.com/t/simple-way-for-multiple-y-axes-for-the-same-query-but-with-different-units/109020) myself), stackoverflow questions and the likes (especially about dual y-axis), but none seems to really provide a proper solution.

of course it's possible we've missed these requests, but providing links would help your case here. to my knowledge, all requests for multiple y axes were not related to showing multiple units for the same data but for showing different datasets in different units on the same graph.

**Comment by calestyo** [comment]
can one mark @github-actions as spam?
